drivers/adc/ad4170/ad4170.c

```diff
@@ -0,0 +1,2123 @@
+ * 	CC BB AA FF EE DD 22 11 00
+ * etc.
+ */
+static void _memswap64(void * buf, uint32_t bytes, size_t step)
+{
+	uint8_t * p = buf;
+	uint32_t i, j;
+	uint8_t temp[8];
+	if (step < 2 || step > 8 || bytes < step || bytes % step != 0)
+		return;
+	for(i = 0; i < bytes; i += step) {
+		memcpy(temp, p, step);
+		for(j = step; j > 0; j--) {
+			*p++ = temp[j-1];
+		}
+	}
+}
+
+static inline void _spi_resync(struct ad4170_dev *dev)
+{
+	uint8_t flush_buf[8];
+	memset(flush_buf, 0xff, sizeof(flush_buf));
+	no_os_spi_write_and_read(dev->spi_desc, flush_buf, sizeof(flush_buf));
+}
+
+static int32_t _verify_sync_pattern(struct ad4170_dev *dev, uint8_t *buf)
+{
+	uint16_t sync_pattern;
+
+	if (!dev->spi_settings.sync_loss_detect ||
+	    dev->config.adc_ctrl.cont_read == AD4170_CONT_TRANSMIT_ON)
+		return 0;
+
+	sync_pattern = (uint16_t)(buf[0] & 0x7F) << 8;
+	if (dev->spi_settings.short_instruction)
+		sync_pattern |= (uint8_t)AD4170_SPI_SYNC_PATTERN;
+	else
+		sync_pattern |= buf[1];
+
+	if (sync_pattern != AD4170_SPI_SYNC_PATTERN) {
+		_spi_resync(dev);
+		return -ECOMM;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Read device register.
+ * @param dev - The device structure.
+ * @param reg_addr - The register address.
+ * @param reg_data - The data read from the register.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad4170_spi_reg_read(struct ad4170_dev *dev,
+			    uint32_t reg_addr,
+			    uint32_t *reg_data)
+{
+	uint8_t buf[7];
+	uint8_t icrc = 0;
+	uint8_t ocrc = 0;
+	uint16_t sz;
+	uint16_t i = 0;
+	uint16_t di;
+	uint8_t reg_size = AD4170_TRANSF_LEN(reg_addr);
+	uint32_t addr = AD4170_ADDR(reg_addr) + reg_size - 1;
+	int32_t ret;
+
+	if (!dev || !reg_data || !_is_addr_valid(dev, reg_addr))
+		return -EINVAL;
+
+	if (dev->config.adc_ctrl.cont_read)
+		return -EACCES;
+
+	memset(buf, 0, sizeof(buf));
+	if (dev->spi_settings.short_instruction) {
+		buf[i++] = (uint8_t)(AD4170_REG_READ_6(addr));
+	} else {
+		buf[i++] = (uint8_t)(AD4170_REG_READ_14(addr) >> 8);
+		buf[i++] = (uint8_t)(addr);
+	}
+
+	di = i; // save the data index
+
+	if (dev->spi_settings.crc_enabled) {
+		ocrc = no_os_crc8(ad4170_crc8, buf, i,
+				  AD4170_CRC8_INITIAL_VALUE); // compute instruction crc
+		icrc = ocrc;
+	}
+
+	i += reg_size; // padding bytes on MOSI (zeroes, due to memset above)
+	sz = i;
+
+	if (dev->spi_settings.crc_enabled) {
+		ocrc = no_os_crc8(ad4170_crc8, &buf[di], reg_size, ocrc);
+		buf[i++] = ocrc;
+		sz = i;
+	}
+
+	ret = no_os_spi_write_and_read(dev->spi_desc, buf, sz);
+	if (ret)
+		return ret;
+
+	ret = _verify_sync_pattern(dev, buf);
+	if (ret < 0)
+		return ret;
+
+	if (dev->spi_settings.crc_enabled) {
+		icrc = no_os_crc8(ad4170_crc8, &buf[di], reg_size, icrc);
+		if (icrc != buf[sz-1])
+			return -EBADMSG;
+	}
+
+	*reg_data = 0;
+	if (dev->big_endian)
+		memcpy((uint8_t *)(reg_data + 1) - reg_size, &buf[di], reg_size);
+	else {
+		memcpy(reg_data, &buf[di], reg_size);
+		_memswap64(reg_data, reg_size, reg_size);
+	}
```

Comment:
Can be replaced by `*reg_data = no_os_get_unaligned_be32()`? The endianess shouldn't matter in that case.

Reply:
From memory, CPU endianess could be little or big, complicated by the fact that the data can come LSB or MSB first on the digital interface and this incoming data may not be aligned to powers of two. memswap64 has a step parameter which could be 3, as in the example. So I'm not touching this part, I'm pretty sure it's not possible to rewrite it using those macros.